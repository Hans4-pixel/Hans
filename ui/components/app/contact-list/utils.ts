export const hasDuplicateContacts = (
  addressBook: any[],
  internalAccounts: any[],
) => {
  const uniqueContactNames = Array.from(
    new Set(addressBook.map(({ name }) => name.toLowerCase().trim())),
  );

  const hasAccountNameCollision = internalAccounts.some((account) =>
    uniqueContactNames.includes(account.metadata.name.toLowerCase().trim()),
  );

  return (
    uniqueContactNames.length !== addressBook.length || hasAccountNameCollision
  );
};

export const isDuplicateContact = (
  addressBook: any[],
  internalAccounts: any[],
  newName: string,
) => {
  const nameExistsInAddressBook = addressBook.some(
    ({ name }) => name.toLowerCase().trim() === newName.toLowerCase().trim(),
  );

  const nameExistsInAccountList = internalAccounts.some(
    ({ metadata }) =>
      metadata.name.toLowerCase().trim() === newName.toLowerCase().trim(),
  );

  return !nameExistsInAddressBook && !nameExistsInAccountList;
};
