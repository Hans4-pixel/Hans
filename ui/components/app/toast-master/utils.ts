import { PayloadAction } from '@reduxjs/toolkit';
import { ReactFragment } from 'react';
import { useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import { DEFAULT_ROUTE } from '../../../helpers/constants/routes';
import { getSwitchedNetworkDetails } from '../../../selectors';
import { SHOW_NFT_DETECTION_ENABLEMENT_TOAST } from '../../../store/actionConstants';
import { submitRequestToBackground } from '../../../store/background-connection';
import { selectSwitchedNetworkNeverShowMessage } from './selectors';

/**
 * Returns true if the privacy policy toast was shown either never, or less than a day ago.
 *
 * @param newPrivacyPolicyToastShownDate
 * @returns true if the privacy policy toast was shown either never, or less than a day ago
 */
export function getIsPrivacyToastRecent(
  newPrivacyPolicyToastShownDate?: number,
): boolean {
  if (!newPrivacyPolicyToastShownDate) {
    return true;
  }

  const currentDate = new Date();
  const oneDayInMilliseconds = 24 * 60 * 60 * 1000;
  const newPrivacyPolicyToastShownDateObj = new Date(
    newPrivacyPolicyToastShownDate,
  );
  const toastWasShownLessThanADayAgo =
    currentDate.valueOf() - newPrivacyPolicyToastShownDateObj.valueOf() <
    oneDayInMilliseconds;

  return toastWasShownLessThanADayAgo;
}

export function setNewPrivacyPolicyToastShownDate(time: number) {
  submitRequestToBackground('setNewPrivacyPolicyToastShownDate', [time]);
}

export function setNewPrivacyPolicyToastClickedOrClosed() {
  submitRequestToBackground('setNewPrivacyPolicyToastClickedOrClosed');
}

export function setShowNftDetectionEnablementToast(
  value: boolean,
): PayloadAction<string | ReactFragment | undefined> {
  return {
    type: SHOW_NFT_DETECTION_ENABLEMENT_TOAST,
    payload: value,
  };
}

export function setSwitchedNetworkNeverShowMessage() {
  submitRequestToBackground('setSwitchedNetworkNeverShowMessage', [true]);
}

export function onHomeScreen() {
  const location = useLocation();

  return location.pathname === DEFAULT_ROUTE;
}

export function setSurveyLinkLastClickedOrClosed(time: number) {
  return async () => {
    await submitRequestToBackground('setSurveyLinkLastClickedOrClosed', [time]);
  };
}

export function getShowAutoNetworkSwitchTest() {
  const switchedNetworkDetails = useSelector(getSwitchedNetworkDetails);
  const switchedNetworkNeverShowMessage = useSelector(
    selectSwitchedNetworkNeverShowMessage,
  );

  return switchedNetworkDetails && !switchedNetworkNeverShowMessage;
}
