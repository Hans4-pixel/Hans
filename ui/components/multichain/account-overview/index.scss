.account-overview-unknown {
  &__empty {
    margin-top: 24px;
  }
}