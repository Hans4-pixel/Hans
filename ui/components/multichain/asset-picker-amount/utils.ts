import { createSelector } from 'reselect';

function getSendInputCurrencySwitched(state) {
  return state.appState.sendInputCurrencySwitched;
}

export const getIsFiatPrimary = createSelector(
  getSendInputCurrencySwitched,
  (sendInputCurrencySwitched) => {
    const isFiatPrimary = Boolean(sendInputCurrencySwitched);
    return isFiatPrimary;
  },
);
