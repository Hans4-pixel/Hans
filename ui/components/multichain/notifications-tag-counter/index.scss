.notifications-tag-counter {
  padding-left: 6px;
  padding-right: 6px;
}

.notifications-tag-counter__unread-dot {
  font-size: 10px;
  padding-left: 3px;
  padding-right: 3px;
  line-height: 1.4;
}