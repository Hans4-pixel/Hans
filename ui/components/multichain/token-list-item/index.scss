.multichain-token-list-item {
  &__container-cell {
    &--clickable {
      &:hover,
      &:focus-within {
        background-color: var(--color-background-default-hover);
        .mm-avatar-network {
            outline-color: var(--color-background-default-hover);
        }
      }
    }
  }

  &__badge {
    align-self: center;
  }
}
