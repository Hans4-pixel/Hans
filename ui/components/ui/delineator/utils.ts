import React from 'react';
import {
  BackgroundColor,
  IconColor,
  TextColor,
} from '../../../helpers/constants/design-system';
import { AvatarIconSize, IconProps, Text } from '../../component-library';
import { DelineatorType } from './delineator.types';

const defaultIconProps = {
  size: AvatarIconSize.Xs,
};

/*
 * Get the icon props based on the type of delineator.
 */
export const getIconPropsByType = (type?: DelineatorType) => {
  let iconProps = {};
  const inverseIconColorProp = {
    color: IconColor.infoInverse,
  } as IconProps<'span'>;

  switch (type) {
    case DelineatorType.Error:
      iconProps = {
        iconProps: inverseIconColorProp,
        backgroundColor: BackgroundColor.errorDefault,
      };
      break;
    default:
      iconProps = {
        iconProps: inverseIconColorProp,
        backgroundColor: BackgroundColor.overlayAlternative,
      };
  }
  return { ...defaultIconProps, ...iconProps };
};

/*
 * Get the text color based on the type of delineator.
 */
const getTextColorByType = (type?: DelineatorType) => {
  switch (type) {
    case DelineatorType.Error:
      return TextColor.errorDefault;
    default:
      return TextColor.textAlternative;
  }
};

/*
 * Override the color of a Text component based on the type of delineator.
 */
export const overrideTextComponentColorByType = ({
  component,
  type,
}: {
  component: React.ReactElement<typeof Text>;
  type?: DelineatorType;
}) => {
  const color = getTextColorByType(type);

  return React.cloneElement(component, {
    color,
  } as React.ComponentProps<typeof Text>);
};
