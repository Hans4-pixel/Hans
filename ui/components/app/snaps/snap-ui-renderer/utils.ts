import { JSXElement, GenericSnapElement } from '@metamask/snaps-sdk/jsx';
import { memoize } from 'lodash';
import { sha256 } from '@noble/hashes/sha256';
import { bytesToHex, remove0x, Json } from '@metamask/utils';
import { unescape as unescapeEntities } from 'he';
import { COMPONENT_MAPPING } from './components';

export type MapToTemplateParams = {
  map: Record<string, number>;
  element: JSXElement;
  form?: string;
};

/**
 * Get a truncated version of component children to use in a hash.
 *
 * @param children - A JSON blob describing one or more children of a component.
 * @returns A truncated version of component children to use in a hash.
 */
function getChildrenForHash(children: Json) {
  if (typeof children === 'string') {
    // For the hash we reduce long strings
    return children.slice(0, 5000);
  }

  if (Array.isArray(children)) {
    // For arrays of children we just use the types
    return children.map((child) => ({ type: child?.type ?? null }));
  }

  return children;
}

/**
 * A memoized function for generating a hash that represents a Snap UI component.
 *
 * This can be used to generate React keys for components.
 *
 * @param component - The component.
 * @returns A hash as a string.
 */
const generateHash = memoize((component: JSXElement) => {
  const { type, props } = component as GenericSnapElement;
  const { name } = props;
  const children = getChildrenForHash(props.children);
  return remove0x(
    bytesToHex(
      sha256(
        JSON.stringify({
          type,
          name: name ?? null,
          children,
        }),
      ),
    ),
  );
});

/**
 * Generate a React key to be used for a Snap UI component.
 *
 * This function also handles collisions between duplicate keys.
 *
 * @param map - A map of previously used keys to be used for collision handling.
 * @param component - The component.
 * @returns A key.
 */
function generateKey(map: Record<string, number>, component: JSXElement) {
  const hash = generateHash(component);
  const count = (map[hash] ?? 0) + 1;
  map[hash] = count;
  return `${hash}_${count}`;
}

export const mapToTemplate = (params: MapToTemplateParams) => {
  const { type } = params.element;
  const indexKey = generateKey(params.map, params.element);
  // @ts-expect-error This seems to be compatibility issue between superstruct and this repo.
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const mapped = COMPONENT_MAPPING[type](params as any);
  return { ...mapped, key: indexKey };
};

export const mapTextToTemplate = (
  elements: unknown[],
  params: Pick<MapToTemplateParams, 'map'>,
) =>
  elements.map((element) => {
    // With the introduction of JSX elements here can be strings.
    if (typeof element === 'string') {
      // We unescape HTML entities here, to allow usage of &bull; etc.
      return unescapeEntities(element);
    }

    return mapToTemplate({ ...params, element });
  });
