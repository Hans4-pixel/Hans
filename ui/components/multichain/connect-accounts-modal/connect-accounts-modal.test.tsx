import React from 'react';
import { BtcAccountType } from '@metamask/keyring-api';
import { waitFor } from '@testing-library/react';
// TODO: Remove restricted import
// eslint-disable-next-line import/no-restricted-paths
import messages from '../../../../app/_locales/en/messages.json';
import configureStore from '../../../store/store';
import mockState from '../../../../test/data/mock-state.json';
import { renderWithProvider } from '../../../../test/jest/rendering';
import { createMockInternalAccount } from '../../../../test/jest/mocks';
import { shortenAddress } from '../../../helpers/utils/util';
// TODO: Remove restricted import
// eslint-disable-next-line import/no-restricted-paths
import { normalizeSafeAddress } from '../../../../app/scripts/lib/multichain/address';
import { ConnectAccountsModal } from './connect-accounts-modal';
const mockAccount = createMockInternalAccount();
const mockBtcAccount = createMockInternalAccount({
    name: 'BTC Account',
    address: 'bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq',
    type: BtcAccountType.P2wpkh,
});
const defaultProps = {
    onClose: () => ({}),
    onAccountsUpdate: () => ({}),
    activeTabOrigin: 'metamask.github.io',
};
const render = (props = defaultProps) => {
    const store = configureStore({
        OnboardingController: {
            completedOnboarding: true
        },
        AccountsController: {
            internalAccounts: {
                accounts: {
                    [mockAccount.id]: mockAccount,
                    [mockBtcAccount.id]: mockBtcAccount,
                },
                selectedAccount: mockAccount.id,
            }
        },
        KeyringController: {
            keyrings: [
                {
                    type: 'HD Key Tree',
                    accounts: [mockAccount.address],
                },
                {
                    type: 'Snap Keyring',
                    accounts: [mockBtcAccount.address],
                },
            ]
        },
        AccountTracker: {
            accounts: {
                [mockAccount.address]: {
                    address: mockAccount.address,
                    balance: '0x0',
                },
            }
        },
        MultichainBalancesController: {
            balances: {
                [mockBtcAccount.id]: {},
            }
        }
    });
    return renderWithProvider(<ConnectAccountsModal>{ ...props } /  > , store);
};
describe('Connect More Accounts Modal', () => {
    it('should render correctly', async () => {
        const { baseElement, getByText } = render();
        expect(getByText(messages.connectMoreAccounts.message)).toBeInTheDocument();
        expect(getByText(messages.selectAll.message)).toBeInTheDocument();
        expect(getByText(messages.confirm.message)).toBeInTheDocument();
        await waitFor(() => {
            expect(baseElement).toMatchSnapshot();
        });
    });
    it('should only render EVM accounts', () => {
        const { getAllByTestId } = render();
        const accountAddresses = getAllByTestId('account-list-address');
        expect(accountAddresses).toHaveLength(1);
        expect(accountAddresses[0].textContent).toBe(
        // Account list item normalizes the address prior to displaying it.
        shortenAddress(normalizeSafeAddress(mockAccount.address)));
    });
});
