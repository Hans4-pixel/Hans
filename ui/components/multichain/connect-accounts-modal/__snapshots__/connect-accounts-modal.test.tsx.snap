// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connect More Accounts Modal should render correctly 1`] = `<div />`;
